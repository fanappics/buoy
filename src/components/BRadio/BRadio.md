Example Radio Object

Radios prop is an array of Radio Objects.

| Key      | Required/Optional | Default | Description                              |
| -------- | ----------------- | ------- | ---------------------------------------- |
| id       | Required          |         | Needs to be unique in the radio group.   |
| value    | Required          |         | Value that gets returned if button is selected. |
| checked  | Optional          | false   | Initial state for radio button is selected. |
| disabled | Optional          | false   | Disables radio button.                   |

Radio Button Options

```js

const radios = [
  {id: 1, value:"Option" },
  {id: 2, value:"Selected", checked: true },
  {id: 3, value:"Disabled", disabled: true }
]

<b-radio groupId="1" groupLabel="Radio Options" :radios="radios" />

```

Disabled Radio Group

```js

const radios2 = [
  {id: 1, value:"Option" },
  {id: 2, value:"Selected", checked: true },
  {id: 3, value:"Disabled", disabled: true }
]

<b-radio disabled groupId="2" groupLabel="Radio Options" :radios="radios2" />

```

Longer Label Name

```js

const radios3 = [
  {id: 1, value:"Option" },
  {id: 2, value:"Selected", checked: true },
  {id: 3, value:"Disabled", disabled: true }
]

<b-radio groupId="3" groupLabel="This is an example of a really long group label." :radios="radios3" />

```

Validation

```js

const radios4 = [
  {id: 1, value:"Select Me" }
]

let isValidRadioGroup = false;

<b-validator v-model="isValidRadioGroup" scope="radio-group-4">
  <b-radio groupId="4" groupLabel="Select to enable button" :radios="radios4" validationName="radio" required />
  <b-button :disabled="!isValidRadioGroup">Submit</b-button>
</b-validator>

```