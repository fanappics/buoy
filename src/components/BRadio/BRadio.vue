<template>
  <div class="container" :class="{ disabled: disabled }">
    <div role="radiogroup" :aria-labelledby="groupId">
      <label :id="groupId" class="description">
      {{ groupLabel }}<span v-if="required" aria-hidden="true">*</span>
      </label>
      <div class="button-group input" role="radiogroup">
        <div v-for="(radio, index) in radios" class="button" :key="index">
          <input type="radio" 
            v-model="publicValue" 
            v-validate.initial="index === 0 ? validations : {}"
            v-bind="index === 0 ? validationAttributes : {}"
            :aria-checked="publicValue === radio.id"
            :aria-describedby="errors.any() ? `error-${groupId}` : ''"
            :aria-labelledby="`${groupId} label-${radio.id}`"
            :aria-posinset="index"
            :aria-setsize="radios.length"
            :disabled="radio.disabled || disabled"
            :id="radio.id"
            :name="groupId"
            :value="radio.id"
          />
          <label :id="`label-${radio.id}`" :for="radio.id">
            {{ radio.value }}
          </label>
        </div>
        <div v-if="errors.any() && showErrors" :id="`error-${groupId}`" class="error-text">
          <span v-for="(error,index) in errors.all()" :key="index">
            {{ error }}
          </span>
        </div>    
      </div>
    </div>
  </div>
</template>

<script>
import validationMixIn from '../../mixins/validation'

export default {
  name: 'b-radio',
  mixins: [validationMixIn()],
  props: {
    //Required props
    /**
    * Each radio group needs a unique id.
    */
    groupId: {
      type: String,
      required: true
    },
    /**
    * Descriptive text about the radio group.
    */
    groupLabel: {
      type: String,
      required: true
    },
    /**
    * Array of Objects containing all buttons included in radio group.
    */
    radios: {
      type: Array,
      required: true,
      validator (value) {
        for (let radio of value) {
          if (!radio.id || !radio.value)
            throw new TypeError('Each radio must have an id and value defined')
        }
        return true
      }
    },
    //Optional props
    /**
    * Disables the entire radio group.
    */
    disabled: Boolean,
    /**
    * Requires a selection to pass validatin. 
    */
    required: Boolean,
    value: String
  },
  data () {
    return {
      showErrors: false,
      checked: this.value || ''
    }
  },
  computed: {
    //Wrapper around value so it is propegated through v-model, or 'public' as I've dubbed it
    publicValue: {
      get () {
        return this.checked
      },
      set (value) {
        this.checked = value.toString()
        this.$emit('input', this.checked)
      }
    },
    validations () {
      let vals = {}
      if (this.required)
        vals.required = true
      return vals
    }
  },
  created: function() {
    const duplicateId = false;
    
    for (let i = 0; i < this.radios.length; i++) {
      if (this.groupId == this.radios[i].id) {
        throw new Error(`Radio option id ${this.radios[i].id} cannot match GroupId`)
      } 
    }
  },
  methods: {
    onValidate (scope) {
      if (!this.parentScope || this.parentScope === scope) {
        this.showErrors = true
      }
    }
  }
}
</script>

<style scoped>
  label {
    flex-grow: 0;
  }

  .button {
    display: flex;
    align-items: baseline;
    padding: .5rem;
  }

  .button > input {
    margin-right: .5rem;
  }

  .button input[type="radio"]:checked+label {
    font-weight: 600;  
  }

  .button-group {
    display: flex;
    flex-direction: column;
    flex: 1;
  }

  .container {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }

  .description {
    font-weight: 600;
  }

</style>
