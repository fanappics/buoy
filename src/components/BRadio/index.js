import BRadio from './BRadio.vue'

/* istanbul ignore next */
BRadio.install = function install (Vue) {
  Vue.component(BRadio.name, BRadio)
}

export default BRadio
