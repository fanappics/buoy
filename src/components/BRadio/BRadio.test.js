import { shallow, createLocalVue } from 'vue-test-utils'
import VeeValidate from 'vee-validate'
import BRadio from './BRadio'

const localVue = createLocalVue()
localVue.use(VeeValidate)

describe('Single radio', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio', value: 'Radio'}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple radio', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio-1', value: 'Radio'}, {id: 'radio-2', value: 'Radio2'}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple radio with default', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio-1', value: 'Radio', checked: true}, {id: 'radio-2', value: 'Radio2'}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple radio with disabled', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio-1', value: 'Radio'}, {id: 'radio-2', value: 'Radio2', disabled: true}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple radio with all disabled', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio-1', value: 'Radio'}, {id: 'radio-2', value: 'Radio2'}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group',
        disabled: true
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple radio with required', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BRadio, {
      propsData: {
        radios: [{id: 'radio-1', value: 'Radio'}, {id: 'radio-2', value: 'Radio2'}],
        groupId: 'radio-group',
        groupLabel: 'Radio Group',
        required: true
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
