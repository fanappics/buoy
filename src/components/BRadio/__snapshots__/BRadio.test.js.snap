// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple radio mounted html should match snapshot 1`] = `
<div class="container">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group
            <!---->
        </label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-1" aria-posinset="0" aria-setsize="2" id="radio-1" name="radio-group" data-vv-as="" data-vv-name="2" data-vv-validate-on="input|blur" data-vv-id="2" aria-required="false"
                    aria-invalid="false">
                <label id="label-radio-1" for="radio-1">
                    Radio
                </label>
            </div>
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-2" aria-posinset="1" aria-setsize="2" id="radio-2" name="radio-group" data-vv-id="3" aria-required="false" aria-invalid="false">
                <label id="label-radio-2" for="radio-2">
                    Radio2
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;

exports[`Multiple radio with all disabled mounted html should match snapshot 1`] = `
<div class="container disabled">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group
            <!---->
        </label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-1" aria-posinset="0" aria-setsize="2" disabled="disabled" id="radio-1" name="radio-group" data-vv-as="" data-vv-name="5" data-vv-validate-on="input|blur" data-vv-id="8"
                    aria-required="false" aria-invalid="false">
                <label id="label-radio-1" for="radio-1">
                    Radio
                </label>
            </div>
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-2" aria-posinset="1" aria-setsize="2" disabled="disabled" id="radio-2" name="radio-group" data-vv-id="9" aria-required="false" aria-invalid="false">
                <label id="label-radio-2" for="radio-2">
                    Radio2
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;

exports[`Multiple radio with default mounted html should match snapshot 1`] = `
<div class="container">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group
            <!---->
        </label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-checked="true" aria-describedby="" aria-labelledby="radio-group label-radio-1" aria-posinset="0" aria-setsize="2" id="radio-1" name="radio-group" data-vv-as="" data-vv-name="3" data-vv-validate-on="input|blur" data-vv-id="4"
                    aria-required="false" aria-invalid="false">
                <label id="label-radio-1" for="radio-1">
                    Radio
                </label>
            </div>
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-2" aria-posinset="1" aria-setsize="2" id="radio-2" name="radio-group" data-vv-id="5" aria-required="false" aria-invalid="false">
                <label id="label-radio-2" for="radio-2">
                    Radio2
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;

exports[`Multiple radio with disabled mounted html should match snapshot 1`] = `
<div class="container">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group
            <!---->
        </label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-1" aria-posinset="0" aria-setsize="2" id="radio-1" name="radio-group" data-vv-as="" data-vv-name="4" data-vv-validate-on="input|blur" data-vv-id="6" aria-required="false"
                    aria-invalid="false">
                <label id="label-radio-1" for="radio-1">
                    Radio
                </label>
            </div>
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-2" aria-posinset="1" aria-setsize="2" disabled="disabled" id="radio-2" name="radio-group" data-vv-id="7" aria-required="false" aria-invalid="false">
                <label id="label-radio-2" for="radio-2">
                    Radio2
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;

exports[`Multiple radio with required mounted html should match snapshot 1`] = `
<div class="container">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group<span aria-hidden="true">*</span></label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-1" aria-posinset="0" aria-setsize="2" id="radio-1" name="radio-group" data-vv-as="" data-vv-name="6" data-vv-validate-on="input|blur" data-vv-id="10" aria-required="true"
                    aria-invalid="false">
                <label id="label-radio-1" for="radio-1">
                    Radio
                </label>
            </div>
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio-2" aria-posinset="1" aria-setsize="2" id="radio-2" name="radio-group" data-vv-id="11" aria-required="false" aria-invalid="false">
                <label id="label-radio-2" for="radio-2">
                    Radio2
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;

exports[`Single radio mounted html should match snapshot 1`] = `
<div class="container">
    <div role="radiogroup" aria-labelledby="radio-group">
        <label id="radio-group" class="description">
            Radio Group
            <!---->
        </label>
        <div role="radiogroup" class="button-group">
            <div class="button">
                <input type="radio" aria-describedby="" aria-labelledby="radio-group label-radio" aria-posinset="0" aria-setsize="1" id="radio" name="radio-group" data-vv-as="" data-vv-name="1" data-vv-validate-on="input|blur" data-vv-id="1" aria-required="false"
                    aria-invalid="false">
                <label id="label-radio" for="radio">
                    Radio
                </label>
            </div>
            <!---->
        </div>
    </div>
</div>
`;
